camel/tasks/task.py

```diff
@@ -0,0 +1,321 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+import re
+from enum import Enum
+from typing import Callable, Dict, List, Literal, Optional, Union
+
+from pydantic import BaseModel
+
+from camel.agents import ChatAgent
+from camel.messages import BaseMessage
+from camel.prompts import TextPrompt
+
+from .task_prompt import TASK_DECOMPOSE_PROMPT, TASK_EVOLVE_PROMPT
+
+
+class TaskSate(str, Enum):
+    OPEN = "OPEN"
+    RUNNING = "RUNNING"
+    DONE = "DONE"
+    DELETED = "DELETED"
+
+    @classmethod
+    def states(cls):
+        return [s.value for s in cls]
+
+
+class Task(BaseModel):
+    """
+    Task is specific assignment that can be passed to a agent.
+
+    Attributes:
+        content: string content for task.
+        id: A unique id.
+        state: The state which should be OPEN, RUNNING, DONE or DELETED.
+        type: task type
+        parent: The parent task, None for root task.
+        subtasks: The childrent sub-tasks for the task.
+        result: The answer for the task.
+    """
+
+    content: str
+    """The content of the task."""
+
+    id: str
+    """An unique string identifier for the task. This should ideally be
+    provided by the provider/model which created the task."""
+
+    state: TaskSate = TaskSate.OPEN
+    """The task state.
+    """
+
+    type: Optional[str] = None
+    """The type of a task.
+    """
+
+    parent: Optional["Task"] = None
+    """The parent task.
+    """
+
+    subtasks: List["Task"] = []
+    """A list of sub tasks.
+    """
+
+    result: str = ""
+
+    @classmethod
+    def from_message(cls, message: BaseMessage) -> "Task":
+        r"""Create a task from a message.
+
+        Args:
+            message (BaseMessage): The message to the task.
+
+        Returns:
+            Task
+        """
+        return cls(content=message.content, id="0")
+
+    @staticmethod
+    def to_message():
+        r"""Convert a Task to a Message."""
+        # TODO
+        pass
+
+    def reset(self):
+        r"""Reset Task to initial state."""
+        self.state = TaskSate.OPEN
+        self.result = ""
+
+    def update_result(self, result: str):
+        self.result = result
+        self.set_state(TaskSate.DONE)
+
+    def set_id(self, id: str):
+        self.id = id
+
+    def set_state(self, state: TaskSate):
+        self.state = state
+        if state == TaskSate.DONE:
+            for subtask in self.subtasks:
+                if subtask.state != TaskSate.DELETED:
+                    subtask.set_state(state)
+        elif state == TaskSate.RUNNING:
+            if self.parent is not None:
+                self.parent.set_state(state)
+
+    def add_subtask(self, task: "Task"):
+        task.parent = self
+        self.subtasks.append(task)
+
```

Comment:
add function to remove subtask?

```suggestion
    def remove_subtask(self, task_id: UUID4):
        if not self.exist(task_id):
            raise ValueError(f"Task ID {task_id} does not exist.")
        self.subtasks = [task for task in self.subtasks if task.id != task_id]
```